web3/middleware/filter.py

```diff
@@ -0,0 +1,324 @@
+from eth_utils import (
+    apply_key_map,
+    to_hex,
+    to_list,
+)
+from toolz.itertoolz import (
+    concat,
+)
+
+
+def segment_count(start, stop=None, step=5):
+    """Creates a segment counting generator
+
+    The generator returns tuple pairs of integers
+    that correspond to segments in the provided range.
+
+    :param start: The initial value of the counting range
+    :param stop: Optional, if provided the last value in the
+    counting range
+    :param step: Optional, the segment length. Default is 5.
+    :type start: int
+    :type stop: int
+    :return: returns a generator object
+
+
+    Example:
+
+    >>> segment_counter = segment_count(start=0, stop=10, step=3)
+    >>> next(segment_counter)
+    (0, 3)
+    >>> next(segment_counter)
+    (3, 6)
+    >>> next(segment_counter)
+    (6, 9)
+    >>> next(segment_counter) #  Remainder is also returned
+    (9, 10)
+    """
+    def finite(start, stop, step):
+        #  If the initial range is less than the step
+        #  just return (start, stop)
+        if stop < step:
+            yield (start, stop)
+            return
+        for i in zip(
+                range(start, stop - step, step),
+                range(start + step, stop, step)):
+            yield i
+        else:
+            remainder = (stop - start) % step
+            #  Handle the remainder
+            if remainder:
+                yield (stop - remainder, stop)
+
+    def infinite(start, step):
+        for i in zip(
+                count(start, step),
+                count(start + step, step)):
+            yield i
+
+    if stop is None:
+        return infinite(start, step)
+    else:
+        return finite(start, stop, step)
+
+
+def block_range_chunker(start_block, last_block=None, step=5):
+    """Returns 2-tuple ranges describing chunks of block from start_block to last_block
+
+       Segments do not overlap to facilitate use as ``toBlock``, ``fromBlock``
+       json-rpc arguments, which are both inclusive.
+    """
+
+    if last_block is not None and start_block > last_block:
+        raise TypeError(
+            "Incompatible start and stop arguments.",
+            "Start must be less than or equal to stop.")
+
+    def block_range_chunker(start_block, last_block, step):
```

Comment:
Generally good to not have this have the same name as the outer function (just for naming clarity).  Also,  I think you can accomplish this without the inner function using:

```
def outer_iter():
    if thing:
        raise Exception(...)
    return (value for value in iter_something())
```

This will hit the validation step when called (as opposed to when `next` is first called.  Do you think readability seems better with that pattern?